File: src/cluster/multiplex.ts
/**
 * Shared request/subscription demultiplexing machinery, mirroring
 * crates/openengine-cluster-client/src/{multiplex.rs,ndjson_pump.rs}. A single
 * {@link MultiplexedTransport} is driven by exactly one wire binding
 * ({@link ../cluster/websocket-transport.js}) in this package, but is itself written generically
 * against a {@link FrameSink} plus fed frames so an alternate binding could reuse it unchanged.
 */

import type { RequestId } from './generated/wire-types.js';
import { getString, isRecord } from './json-guards.js';
import {
  BoundedChannel,
  extractRequestId,
  JSON_RPC_VERSION,
  TransportError,
  type JsonRpcNotificationEnvelope,
  type OverflowFlag,
  type PumpedSubscription,
  type SubscriptionTransport,
} from './transport.js';

/** Bounded per-subscription local buffer of raw notification lines awaiting delivery, matching
 * the server's `serve_ndjson`/`serve_websocket` bound in crates/openengine-cluster-client/src/
 * lib.rs (`SUBSCRIPTION_QUEUE_CAPACITY`). */
export const SUBSCRIPTION_QUEUE_CAPACITY = 1024;

/** Abstraction over "write one already-serialized JSON-RPC frame to the peer" -- implemented
 * once per wire transport (only {@link ../cluster/websocket-transport.js WebSocketFrameSink} in
 * this package) so the demultiplexing logic below is implemented exactly once regardless of
 * frame shape. Mirrors `multiplex::FrameSink`. */
export interface FrameSink {
  sendFrame(frame: string): Promise<void>;
}

interface Deferred<T> {
  resolve(value: T): void;
  reject(error: unknown): void;
}

function createDeferred<T>(): { promise: Promise<T> } & Deferred<T> {
  let resolve!: (value: T) => void;
  let reject!: (error: unknown) => void;
  const promise = new Promise<T>((res, rej) => {
    resolve = res;
    reject = rej;
  });
  return { promise, resolve, reject };
}

interface PumpedResponse {
  line: string;
  subscription?: PumpedSubscription;
}

interface SubscriptionRegistration {
  channel: BoundedChannel<string>;
  overflowed: OverflowFlag;
}

type PendingMap = Map<RequestId, Deferred<PumpedResponse>>;
type SubscriptionMap = Map<string, SubscriptionRegistration>;

/** Best-effort coercion of an unknown `id` field into a {@link RequestId}, mirroring
 * `RequestId::from_json_value`: only a string or a finite number counts as a resolvable id;
 * anything else (missing, `null`, object, array) yields `undefined` so the caller drops the
 * frame instead of throwing on peer-controlled input. */
// eslint-disable-next-line sonarjs/function-return-type -- RequestId is `string | number` by design (mirrors Rust's `enum RequestId`).
function coerceResponseId(id: unknown): RequestId | undefined {
  if (typeof id === 'string') {
    return id;
  }
  if (typeof id === 'number' && Number.isFinite(id)) {
    return id;
  }
  return undefined;
}

/** Forwards one `event`/`subscription/closed` notification without waiting on a consumer.
 * Returns the subscription id when the local channel is full and the server must be cancelled.
 * Mirrors `ndjson_pump::forward_notification`. */
function forwardNotification(
  value: Record<string, unknown>,
  line: string,
  subscriptions: SubscriptionMap
): string | undefined {
  const params = isRecord(value.params) ? value.params : undefined;
  const subscriptionId = params ? getString(params, 'subscriptionId') : undefined;
  if (subscriptionId === undefined) {
    return undefined;
  }
  const terminal = value.method === 'subscription/closed';
  const registration = subscriptions.get(subscriptionId);
  if (!registration) {
    return undefined;
  }
  if (registration.channel.trySend(line)) {
    if (terminal) {
      subscriptions.delete(subscriptionId);
    }
    return undefined;
  }
  registration.overflowed.value = true;
  subscriptions.delete(subscriptionId);
  return terminal ? undefined : subscriptionId;
}

/** Decodes and routes one pumped frame: a notification is forwarded live (see
 * {@link forwardNotification}); a unary response resolves its pending deferred, registering a
 * freshly minted subscription's channel first when the response is a successful
 * subscription-establishing result carrying `result.subscriptionId` -- so no `event` racing the
 * response can be missed. Malformed JSON, a notification/response with no resolvable identity, or
 * an unknown/already-resolved request id are silently dropped. Mirrors
 * `ndjson_pump::route_pumped_message`. */
function routePumpedMessage(
  line: string,
  pending: PendingMap,
  subscriptions: SubscriptionMap
): string | undefined {
  let value: unknown;
  try {
    value = JSON.parse(line);
  } catch {
    return undefined;
  }
  if (!isRecord(value)) {
    return undefined;
  }
  if (typeof value.method === 'string') {
    return forwardNotification(value, line, subscriptions);
  }
  const id = coerceResponseId(value.id);
  if (id === undefined) {
    return undefined;
  }
  const deferred = pending.get(id);
  if (!deferred) {
    return undefined;
  }
  pending.delete(id);

  const result = isRecord(value.result) ? value.result : undefined;
  const resultSubscriptionId = result ? getString(result, 'subscriptionId') : undefined;
  let subscription: PumpedSubscription | undefined;
  if (resultSubscriptionId !== undefined) {
    const channel = new BoundedChannel<string>(SUBSCRIPTION_QUEUE_CAPACITY);
    const overflowed: OverflowFlag = { value: false };
    subscriptions.set(resultSubscriptionId, { channel, overflowed });
    subscription = { channel, overflowed };
  }
  deferred.resolve(subscription === undefined ? { line } : { line, subscription });
  return undefined;
}

/** Fails every still-pending request and ends every open subscription channel once the pump's
 * read half ends -- mirrors `multiplex::finish_pump`. */
function finishPump(pending: PendingMap, subscriptions: SubscriptionMap): void {
  for (const deferred of pending.values()) {
    deferred.reject(new TransportError('server closed the connection before responding'));
  }
  pending.clear();
  for (const registration of subscriptions.values()) {
    registration.channel.close();
  }
  subscriptions.clear();
}

/** Registers `id` as pending, writes `request`, and awaits its demultiplexed response. Mirrors
 * `multiplex::send_request`. */
async function sendRequest(
  sink: FrameSink,
  pending: PendingMap,
  request: string,
  id: RequestId
): Promise<PumpedResponse> {
  if (pending.has(id)) {
    throw new TransportError(`request id is already pending: ${JSON.stringify(id)}`);
  }
  const deferred = createDeferred<PumpedResponse>();
  pending.set(id, deferred);
  try {
    await sink.sendFrame(request);
  } catch (error) {
    pending.delete(id);
    throw error instanceof TransportError ? error : new TransportError(String(error), { cause: error });
  }
  return deferred.promise;
}

async function cancelSubscriptionFrame(sink: FrameSink, subscriptionId: string): Promise<void> {
  const notification: JsonRpcNotificationEnvelope<'subscription/cancel', { subscriptionId: string }> = {
    jsonrpc: JSON_RPC_VERSION,
    method: 'subscription/cancel',
    params: { subscriptionId },
  };
  await sink.sendFrame(JSON.stringify(notification));
}

async function cancelRequestFrame(sink: FrameSink, id: RequestId): Promise<void> {
  const notification: JsonRpcNotificationEnvelope<'$/cancelRequest', { id: RequestId }> = {
    jsonrpc: JSON_RPC_VERSION,
    method: '$/cancelRequest',
    params: { id },
  };
  await sink.sendFrame(JSON.stringify(notification));
}

/** Routes one decoded frame and, if it named a subscription whose local channel has overflowed,
 * best-effort sends its cancellation. Mirrors `multiplex::route_and_maybe_cancel`. */
async function routeAndMaybeCancel(
  line: string,
  pending: PendingMap,
  subscriptions: SubscriptionMap,
  sink: FrameSink
): Promise<void> {
  const subscriptionId = routePumpedMessage(line, pending, subscriptions);
  if (subscriptionId !== undefined) {
    try {
      await cancelSubscriptionFrame(sink, subscriptionId);
    } catch {
      // Best-effort: mirrors `let _ = cancel_subscription(...).await` in multiplex.rs.
    }
  }
}

/**
 * Owns one connection's demultiplexing state -- write sink, pending-request map, subscription
 * registry, and per-connection watch-id counter -- and implements
 * {@link SubscriptionTransport} against it exactly once. A wire binding constructs one of these
 * from its {@link FrameSink} and feeds it every inbound frame via {@link routeIncomingFrame},
 * calling {@link endStream} once when its read half ends. Mirrors `multiplex::
 * MultiplexedTransport`, folded together with the pump-loop wiring `NdjsonTransport`/
 * `WebSocketTransport` each did by hand in Rust, since this package has only one wire binding.
 */
export class MultiplexedTransport implements SubscriptionTransport {
  private readonly sink: FrameSink;
  private readonly pending: PendingMap = new Map();
  private readonly subscriptions: SubscriptionMap = new Map();
  private nextWatchId = 1;

  constructor(sink: FrameSink) {
    this.sink = sink;
  }

  async request(request: string): Promise<string> {
    const id = extractRequestId(request);
    const response = await sendRequest(this.sink, this.pending, request, id);
    return response.line;
  }

  openSubscription(
    request: string,
    id: RequestId
  ): Promise<{ line: string; subscription?: PumpedSubscription }> {
    return sendRequest(this.sink, this.pending, request, id);
  }

  async cancelSubscription(subscriptionId: string): Promise<void> {
    await cancelSubscriptionFrame(this.sink, subscriptionId);
  }

  async cancelRequest(id: RequestId): Promise<void> {
    await cancelRequestFrame(this.sink, id);
  }

  // eslint-disable-next-line sonarjs/function-return-type -- RequestId is `string | number` by design (mirrors Rust's `enum RequestId`).
  nextWatchRequestId(): RequestId {
    return `watch-${this.nextWatchId++}`;
  }

  /** Feeds one inbound frame (a WebSocket text message) to the demultiplexer. Fire-and-forget
   * from the caller's perspective, matching the pump loop's `await` on
   * `route_and_maybe_cancel` in Rust -- ordering across frames is still preserved because this
   * package's WebSocket binding awaits each call before processing the next `message` event. */
  async routeIncomingFrame(line: string): Promise<void> {
    await routeAndMaybeCancel(line, this.pending, this.subscriptions, this.sink);
  }

  /** Ends the pump: fails every pending request and closes every open subscription channel.
   * Mirrors `multiplex::finish_pump`, called once the WebSocket connection closes. */
  endStream(): void {
    finishPump(this.pending, this.subscriptions);
  }
}

File: crates/openengine-cluster-client/src/multiplex.rs
//! Shared request/subscription demultiplexing machinery, implemented once and reused verbatim by
//! [`crate::NdjsonTransport`] (NDJSON lines) and [`crate::websocket::WebSocketTransport`]
//! (`Message::Text` frames) so both wire bindings drive this crate's [`crate::JsonRpcTransport`]/
//! [`crate::SubscriptionTransport`] methods through identical code regardless of frame shape.

use std::collections::hash_map::Entry;
use std::sync::atomic::{AtomicU64, Ordering};

use async_trait::async_trait;
use openengine_cluster_protocol::{
    CancelRequestParams, JsonRpcNotification, RequestId, SubscriptionCancelParams, SubscriptionId,
    JSON_RPC_VERSION,
};
use tokio::sync::oneshot;
use tokio::task::JoinHandle;

use crate::ndjson_pump::route_pumped_message;
use crate::{
    extract_request_id, JsonRpcTransport, PendingMap, PumpedResponse, PumpedSubscription,
    SubscriptionMap, SubscriptionTransport, TransportError,
};

/// Abstraction over "write one already-serialized JSON-RPC frame to the peer", implemented once
/// per wire transport ([`crate::NdjsonFrameSink`], [`crate::websocket::WebSocketFrameSink`]) so
/// the demultiplexing logic below is implemented exactly once regardless of the underlying frame
/// shape (NDJSON line vs. `Message::Text`).
#[async_trait]
pub(crate) trait FrameSink: Send + Sync {
    async fn send_frame(&self, frame: String) -> Result<(), TransportError>;
}

/// Registers `id` as pending, writes `request`, and awaits its demultiplexed response. Shared body
/// of `NdjsonTransport::send_request` and `WebSocketTransport::send_request`.
pub(crate) async fn send_request<F: FrameSink>(
    sink: &F,
    pending: &PendingMap,
    request: String,
    id: RequestId,
) -> Result<PumpedResponse, TransportError> {
    let (sender, receiver) = oneshot::channel();
    {
        let mut pending = pending.lock();
        match pending.entry(id.clone()) {
            Entry::Vacant(entry) => {
                entry.insert(sender);
            }
            Entry::Occupied(_) => {
                return Err(TransportError::Protocol(format!(
                    "request id is already pending: {id:?}"
                )));
            }
        }
    }
    if let Err(error) = sink.send_frame(request).await {
        pending.lock().remove(&id);
        return Err(error);
    }
    receiver.await.map_err(|_| {
        TransportError::Protocol("server closed the connection before responding".to_owned())
    })
}

/// Shared body of both transports' [`crate::JsonRpcTransport::request`] impl.
pub(crate) async fn request<F: FrameSink>(
    sink: &F,
    pending: &PendingMap,
    request: String,
) -> Result<String, TransportError> {
    let id = extract_request_id(&request)?;
    Ok(send_request(sink, pending, request, id).await?.line)
}

/// Shared body of both transports' [`crate::SubscriptionTransport::open_subscription`] impl.
pub(crate) async fn open_subscription<F: FrameSink>(
    sink: &F,
    pending: &PendingMap,
    request: String,
    id: RequestId,
) -> Result<(String, Option<PumpedSubscription>), TransportError> {
    let response = send_request(sink, pending, request, id).await?;
    Ok((response.line, response.subscription))
}

/// Shared body of both transports' [`crate::SubscriptionTransport::cancel_subscription`] impl.
pub(crate) async fn cancel_subscription<F: FrameSink>(
    sink: &F,
    subscription_id: SubscriptionId,
) -> Result<(), TransportError> {
    let notification = serde_json::to_string(&JsonRpcNotification {
        jsonrpc: JSON_RPC_VERSION.to_owned(),
        method: "subscription/cancel".to_owned(),
        params: SubscriptionCancelParams { subscription_id },
    })
    .map_err(|error| TransportError::Protocol(error.to_string()))?;
    sink.send_frame(notification).await
}

/// Shared body of both transports' [`crate::SubscriptionTransport::cancel_request`] impl.
pub(crate) async fn cancel_request<F: FrameSink>(
    sink: &F,
    id: RequestId,
) -> Result<(), TransportError> {
    let notification = serde_json::to_string(&JsonRpcNotification {
        jsonrpc: JSON_RPC_VERSION.to_owned(),
        method: "$/cancelRequest".to_owned(),
        params: CancelRequestParams { id },
    })
    .map_err(|error| TransportError::Protocol(error.to_string()))?;
    sink.send_frame(notification).await
}

/// Shared body of both transports' [`crate::SubscriptionTransport::next_watch_request_id`] impl.
pub(crate) fn next_watch_id(counter: &AtomicU64) -> RequestId {
    RequestId::String(format!("watch-{}", counter.fetch_add(1, Ordering::Relaxed)))
}

/// Routes one decoded message body via [`route_pumped_message`] and, if it named a subscription
/// whose local queue has overflowed or been abandoned, best-effort sends its cancellation --
/// shared by both transports' pump loops.
pub(crate) async fn route_and_maybe_cancel<F: FrameSink>(
    line: String,
    pending: &PendingMap,
    subscriptions: &SubscriptionMap,
    sink: &F,
) {
    if let Some(subscription_id) = route_pumped_message(line, pending, subscriptions) {
        let _ = cancel_subscription(sink, subscription_id).await;
    }
}

/// Fails every still-pending request and ends every open subscription (dropping its sender) once
/// a pump's read half ends -- shared tail of both transports' pump loops.
pub(crate) fn finish_pump(pending: &PendingMap, subscriptions: &SubscriptionMap) {
    for (_, sender) in pending.lock().drain() {
        drop(sender);
    }
    subscriptions.lock().clear();
}

/// Owns one connection's demultiplexing state -- write sink, pending-request map, read-half pump
/// task, and per-connection watch-id counter -- and implements [`crate::JsonRpcTransport`]/
/// [`crate::SubscriptionTransport`] exactly once against it. [`crate::NdjsonTransport`] and
/// [`crate::websocket::WebSocketTransport`] each hold one of these behind their public,
/// frame-shape-specific type and forward every trait method to it single-line, so the demux
/// wiring -- not just the routing logic in the free functions above -- is implemented once
/// regardless of frame shape. (A blanket `impl<F: FrameSink> JsonRpcTransport for
/// MultiplexedTransport<F>` cannot itself satisfy `NdjsonTransport`/`WebSocketTransport`'s trait
/// bounds without a wrapper: Rust's coherence rules reject a second blanket impl over an
/// unconstrained type parameter alongside the existing `impl<T: JsonRpcTransport + ?Sized>
/// JsonRpcTransport for &T` forwarding impl, since `T` could unify with `&_`.)
pub(crate) struct MultiplexedTransport<F> {
    sink: F,
    pending: PendingMap,
    pump: JoinHandle<()>,
    next_watch_id: AtomicU64,
}

impl<F: FrameSink> MultiplexedTransport<F> {
    pub(crate) fn new(sink: F, pending: PendingMap, pump: JoinHandle<()>) -> Self {
        Self {
            sink,
            pending,
            pump,
            next_watch_id: AtomicU64::new(1),
        }
    }
}

impl<F> Drop for MultiplexedTransport<F> {
    fn drop(&mut self) {
        self.pump.abort();
    }
}

#[async_trait]
impl<F: FrameSink> JsonRpcTransport for MultiplexedTransport<F> {
    async fn request(&self, req: String) -> Result<String, TransportError> {
        request(&self.sink, &self.pending, req).await
    }
}

#[async_trait]
impl<F: FrameSink> SubscriptionTransport for MultiplexedTransport<F> {
    async fn open_subscription(
        &self,
        req: String,
        id: RequestId,
    ) -> Result<(String, Option<PumpedSubscription>), TransportError> {
        open_subscription(&self.sink, &self.pending, req, id).await
    }

    async fn cancel_subscription(
        &self,
        subscription_id: SubscriptionId,
    ) -> Result<(), TransportError> {
        cancel_subscription(&self.sink, subscription_id).await
    }

    async fn cancel_request(&self, id: RequestId) -> Result<(), TransportError> {
        cancel_request(&self.sink, id).await
    }

    fn next_watch_request_id(&self) -> RequestId {
        next_watch_id(&self.next_watch_id)
    }
}

/// Generates `JsonRpcTransport`/`SubscriptionTransport` for a wire-transport wrapper type holding
/// an `inner: MultiplexedTransport<_>` field, forwarding every method to it. Written once here (as
/// a macro, not a blanket impl -- see [`MultiplexedTransport`]'s doc comment for why a blanket impl
/// does not typecheck) so [`crate::NdjsonTransport`] and [`crate::websocket::WebSocketTransport`]
/// each get one macro invocation instead of hand-writing the same forwarding source twice.
macro_rules! impl_multiplexed_transport {
    ($ty:ident < $($generic:ident),+ > where $($bound:tt)+) => {
        #[async_trait::async_trait]
        impl<$($generic),+> crate::JsonRpcTransport for $ty<$($generic),+>
        where
            $($bound)+
        {
            async fn request(&self, request: String) -> Result<String, crate::TransportError> {
                self.inner.request(request).await
            }
        }

        #[async_trait::async_trait]
        impl<$($generic),+> crate::SubscriptionTransport for $ty<$($generic),+>
        where
            $($bound)+
        {
            /// Establishment can legitimately fail with a JSON-RPC error (for example
            /// `agent/attach` rejecting an unknown or inactive `ExecutionRef`) -- that case
            /// carries no `subscriptionId` but is not a transport fault, so it is left for the
            /// caller's response parser to surface as a typed `ClientError::Rpc` rather than
            /// being collapsed into a generic `TransportError` here.
            async fn open_subscription(
                &self,
                request: String,
                id: openengine_cluster_protocol::RequestId,
            ) -> Result<(String, Option<crate::PumpedSubscription>), crate::TransportError> {
                self.inner.open_subscription(request, id).await
            }

            async fn cancel_subscription(
                &self,
                subscription_id: openengine_cluster_protocol::SubscriptionId,
            ) -> Result<(), crate::TransportError> {
                self.inner.cancel_subscription(subscription_id).await
            }

            async fn cancel_request(
                &self,
                id: openengine_cluster_protocol::RequestId,
            ) -> Result<(), crate::TransportError> {
                self.inner.cancel_request(id).await
            }

            fn next_watch_request_id(&self) -> openengine_cluster_protocol::RequestId {
                self.inner.next_watch_request_id()
            }
        }
    };
}

pub(crate) use impl_multiplexed_transport;
